### PR TITLE
New FPGA code sample board_test

### PR DESCRIPTION
**Pull request by aktole**
Signed-off-by: Anuja Ketan Tole <46543365+aktole@users.noreply.github.com>

# Existing Sample Changes
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes Issue# 

## External Dependencies

List any external dependencies created as a result of this change.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode
- [x] When compiling the compliler flag "-Wall -Wformat-security -Werror=format-security" was used

**_Delete this line and everything below if this is not a PR for a new code sample_**

**_Delete this line and all above it if this PR is for a new code sample_**
# Adding a New Sample(s)

## Description

Platform test (board test) added to repository. The test is for platform developers to check board interfaces and measure board metrics like host to device memory bandwidth, kernel to memory bandwidth, kernel clock frequency and kernel launch latency.

## Checklist
Administrative
- [X] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): Yohann Uguen
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [X] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 1]
```diff
@@ -0,0 +1,346 @@
+# Board Test
+This design contains tests to check FPGA board interfaces and reports the following metrics:
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I'm sorry I'll have to make you move this test in "Reference Designs" as this is not demonstrating a tool.
```suggestion
This reference design contains tests to check FPGA board interfaces and reports the following metrics:

```

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | How to test board interfaces to ensure designed platform provides expected performance
+| Time to complete                  | 30 minutes (not including compile time)
+
+## Purpose
+This test can be used to check FPGA board interfaces and measure host to device as well as kernel to global memory interface metrics. Custom platform developers can use this test as a starting point to validate custom platform interfaces.
+
+### Introduction to oneAPI FPGA Board Support Package(BSP)
+
+A oneAPI BSP consists of software layers and an FPGA hardware scaffold design that makes it possible to target the FPGA through the Intel® oneAPI DPC++/C++ Compiler. The FPGA design generated by the compiler is stitched into the framework provided by the oneAPI BSP. Please see information about oneAPI BSPs in [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-bsps-and-boards.html).
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]:
```suggestion
A oneAPI board support package (BSP) consists of software layers and an FPGA hardware scaffold design that makes it possible to target the FPGA through the Intel® oneAPI DPC++/C++ Compiler. The FPGA design generated by the compiler is stitched into the framework provided by the oneAPI BSP. Please see information about oneAPI BSPs in [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-bsps-and-boards.html). 

```

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+### Introduction to oneAPI FPGA Board Support Package(BSP)
+
+A oneAPI BSP consists of software layers and an FPGA hardware scaffold design that makes it possible to target the FPGA through the Intel® oneAPI DPC++/C++ Compiler. The FPGA design generated by the compiler is stitched into the framework provided by the oneAPI BSP. Please see information about oneAPI BSPs in [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/fpga-bsps-and-boards.html). 
+
+The BSP hardware components typically comprise RTL for all interfaces required by the oneAPI kernel, for example a PCIe IP for host to kernel communication, EMIF (External Memory Interface) IP for kernel to memory as well as host to FPGA board memory communication etc.
+
+The BSP software components typically consists of a MMD (Memory Mapped Device) layer and a driver. The implementation is vendor dependent.
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]:
```suggestion
The BSP software components typically consists of a Memory Mapped Device (MMD) layer and a driver. The implementation is vendor dependent.

```

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+The complete test is divided into six sub-tests. By default, all tests are run. You can choose to run a single test using `-test=<test number>` option. Please see [Running the Sample](#running-the-sample) for test usage instructions.
+1. Host Speed and Host Read Write Test
+2. Kernel Clock Frequency Test
+3. Kernel Launch Test
+4. Kernel Latency Measurement
+5. Kernel-to-Memory Read Write Test
+6. Kernel-to-Memory Bandwidth Test<br>
+*Note:* It is recommended to run all tests at least once to ensure that the platform interfaces are fully functional.
+
+## Key Concepts
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I think that you can remove this section as I feel it does not bring much information

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+* Sample to demonstrate platform interface checks
+
+## License  
+Code samples are licensed under the MIT license. See
+[License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.
+
+Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)
+
+## Building the `board_test` code sample
+
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]:
```suggestion

> **Note**: If you have not already done so, set up your CLI
> environment by sourcing  the `setvars` script located in
> the root of your oneAPI installation.
>
> Linux Sudo: . /opt/intel/oneapi/setvars.sh
>
> Linux User: . ~/intel/oneapi/setvars.sh
>
> Windows: C:\Program Files(x86)\Intel\oneAPI\setvars.bat
>
>For more information on environment variables, see Use the setvars Script for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
```

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+### Include Files
+The included header `dpc_common.hpp` is located at `$ONEAPI_ROOT/dev-utilities/latest/include` on your development system.
+
+### Running Samples in DevCloud
+If running a sample in the Intel DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the [Intel® oneAPI Base Toolkit Get Started Guide](https://devcloud.intel.com/oneapi/documentation/base-toolkit/).
+
+When compiling for FPGA hardware, it is recommended to increase the job timeout to 12h.
+
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: If you are planning on adding Windows support, you'll have to add a section on how to compile on Windows

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+     ```
+   * Compile for FPGA hardware (takes longer to compile, targets FPGA device):
+     ```
+     make fpga
+     ```
+3. (Optional) As the above hardware compile may take several hours to complete, FPGA pre-compiled binaries (compatible with Linux* Ubuntu* 18.04) can be downloaded <a href="https://iotdk.intel.com/fpga-precompiled-binaries/latest/board_test.fpga.tar.gz" download>here</a>.
+
+### In Third-Party Integrated Development Environments (IDEs)
+
+You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]:
```suggestion
You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)

```

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+## Examining the Reports
+Locate `report.html` in the `board_test.prj/reports/` directory. Open the report in any of Chrome*, Firefox*, Edge*, or Internet Explorer*.
+
+You may also wish to examine the reports generated by the full FPGA hardware compile and compare their contents.
+
+## Running the Sample
+
+ 1. Run the sample on the FPGA emulator (the kernel executes on the CPU):
+     ```
+     ./board_test.fpga_emu     (Linux)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: If you add Windows support, you'll need to add Windows commands as well

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+     ```
+2. Run the sample on the FPGA device:
+     ```
+     ./board_test.fpga         (Linux)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: If you add Windows support, you'll need to add Windows commands as well

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+     ```
+The above commands run all the test described in [Test Details](#test-details), to run a specific test pass the test number as argument to `-test=<test number>` option: 
+```
+./board_test.fpga -test=<test number>     (Linux)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: If you add Windows support, you'll need to add Windows commands as well

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+*Note:* The test allows overriding this failure, however, this is may lead to functional errors and is not recommended. The override option is provided to allow debug in case where platform design changes are done to force kernel to run at slower clock (this is not a common use-case). To override, set `report_chk` variable to `false` in `board_test.cpp` and recompile host code only using `-reuse-exe=board_test.fpga` option in compile command.
+
+2. Requirement for kernel-to-memory bandwidth test: Export `HLD_SHIM_ROOT_HW` to point to `path-to-oneAPI-shim/hardware/<board_variant>` directory. The test looks for `board_spec.xml` file in this folder to get the global memory information for the board. The test does not fail if this variable is not set, it reports measured kernel to device global memory bandwidth numbers and continues. 
+
+To view test details and usage information using the binary, use `-help` option.
+```
+./board_test.fpga -help      (Linux)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: If you add Windows support, you'll need to add Windows commands as well

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+It is assumed that all memory interfaces have equal widths. 
+
+BOARD BANDWIDTH UTILIZATION = 85.4672
+Warning : Board bandwidth utilization is less than 90% 
+
+KERNEL-TO-MEMORY BANDWIDTH = 16409.7 MB/s/bank
+
+BOARD TEST PASSED
+ 
+```
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: We usually have an additional section for Reference Designs, e.g.:
```
## Additional Design Information
### Source Code Explanation

| File                         | Description
---                            |---
| `gzip.cpp`                   | Contains the `main()` function and the top-level interfaces to the SYCL* GZIP functions.
| `gzip_ll.cpp`                | Low latency variant of the top level file.
| `gzipkernel.cpp`             | Contains the SYCL* kernels used to implement GZIP.
| `gzipkernel_ll.cpp`          | Low-latency variant of kernels.
...

### Compiler Flags Used

| Flag | Description
---    |---
`-Xshardware` | Target FPGA hardware (as opposed to FPGA emulator)
`-Xsparallel=2` | Uses two cores when compiling the bitstream through Quartus®
`-Xsseed=<seed_num>` | Uses a particular seed while running Quartus®, selected to yield the best Fmax for this desgin
`-Xsnum-reorder=6` | On Intel Stratix® 10 SX only, specify a wider data path for read data from global memory
`-Xsopt-arg="-nocaching"` | Specifies that cached LSUs should not be used.
`-DNUM_ENGINES=<1|2>` | Specifies that 1 GZIP engine should be compiled when targeting Intel Arria® 10 GX and two engines when targeting Intel Stratix® 10 SX
```

**Comment by yuguen** [comment]
Sorry I started my review with the wrong code version, you'll have to go over both reviews :(

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 1 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: updated readme

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 4 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: done

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 6 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: Added a section on Windows, will be in next upload

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 8 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I have added Windows section, it will be in next upload

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 9 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I have added Windows section, it will be in next upload

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 10 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I have added Windows section, it will be in next upload

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 11 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: I have added Windows section, it will be in next upload

File: DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md (hunk 12 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/Tutorials/Tools/board_test/README.md]: Added section with source file and compiler flag information

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md [hunk 13]
```diff
@@ -1,74 +1,75 @@
...
-3. Kernel launch latency
-4. Kernel to device global memory bandwidth
+* Host to device global memory interface bandwidth
+* Kernel clock frequency
+* Kernel launch latency
+* Kernel to device global memory bandwidth
 
-***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
-The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
+***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you navigate the samples and build your knowledge on SYCL* for FPGA. <br>
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: We cannot make changes to the Documentation section. Please revert the changes in this section

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md [hunk 14]
```diff
@@ -1,74 +1,75 @@
...
-*Note:* It is recommended to run all tests at least once to ensure that the platform interfaces are fully functional.
+   
+***Note:*** Intel recommends running all tests at least once to ensure that the platform interfaces are fully functional.
 
 ## License  
 Code samples are licensed under the MIT license. See
 [License.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/License.txt) for details.
 
-Third party program Licenses can be found here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)
+You can find third-party program licenses here: [third-party-programs.txt](https://github.com/oneapi-src/oneAPI-samples/blob/master/third-party-programs.txt)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: We cannot make changes to the License section, please revert these changes.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md [hunk 15]
```diff
@@ -87,9 +88,11 @@ Third party program Licenses can be found here: [third-party-programs.txt](https
 The included header `dpc_common.hpp` is located at `$ONEAPI_ROOT/dev-utilities/latest/include` on your development system.
 
 ### Running Samples in DevCloud
-If running a sample in the Intel DevCloud, remember that you must specify the type of compute node and whether to run in batch or interactive mode. Compiles to FPGA are only supported on fpga_compile nodes. Executing programs on FPGA hardware is only supported on fpga_runtime nodes of the appropriate type, such as fpga_runtime:arria10 or fpga_runtime:stratix10.  Neither compiling nor executing programs on FPGA hardware are supported on the login nodes. For more information, see the [Intel® oneAPI Base Toolkit Get Started Guide](https://devcloud.intel.com/oneapi/documentation/base-toolkit/).
+If running a sample in the Intel DevCloud, specify the compute node type and whether to run in batch or interactive mode.
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: We cannot make changes to the "Running Samples in DevCloud" section, please revert these changes.

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md [hunk 16]
```diff
@@ -166,12 +169,12 @@ If an error occurs, you can get more details by running make with the VERBOSE=1
 
 ### In Third-Party Integrated Development Environments (IDEs)
 
-You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)
+You can compile and run this tutorial in the Eclipse* IDE (in Linux*). For instructions, refer to the following link:<br> [FPGA Workflows on Third-Party IDEs for Intel® oneAPI Toolkits](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: We cannot make changes to this section, please revert these changes.

**Review by yuguen — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md (hunk 13 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: done

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md (hunk 14 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: done

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md (hunk 15 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: done

File: DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md (hunk 16 above)
Comment (aktole) [reply DirectProgramming/DPC++FPGA/ReferenceDesigns/board_test/README.md]: done

**Comment by aktole** [comment]
Internal review complete, all requested changes merged.

**Comment by aktole** [comment]
Internal review complete, all requested changes merged.
